backends/ebpf/psa/ebpfPsaTable.cpp

```diff
@@ -564,4 +579,344 @@ bool EBPFTablePSA::dropOnNoMatchingEntryFound() const {
+        builder->endOfStatement(true);
+
+        for (size_t i = 0; i < keyGenerator->keyElements.size(); i++) {
+            auto keyElement = keyGenerator->keyElements[i];
+            auto expr = firstEntry->keys->components[i];
+            cstring fieldName = program->refMap->newName("field");
+            auto ebpfType = get(keyTypes, keyElement);
+            builder->emitIndent();
+            ebpfType->declare(builder, fieldName, false);
+            builder->append(" = ");
+            if (auto mask = expr->to<IR::Mask>()) {
+                mask->right->apply(cg);
+                builder->endOfStatement(true);
+            } else {
+                // MidEnd transforms 0xffff... masks into exact match
+                // So we receive there a Constant same as exact match
+                // So we have to create 0xffff... mask on our own
+                unsigned width = 0;
+                if (auto hasWidth = ebpfType->to<IHasWidth>()) {
+                    width = hasWidth->widthInBits();
+                } else {
+                    BUG("Cannot assess field bit width");
+                }
+                builder->append("0x");
+                for (int j = 0; j < width / 8; j++)
+                    builder->append("ff");
+                builder->endOfStatement(true);
+            }
+            builder->emitIndent();
+            builder->appendFormat("__builtin_memcpy(%s, &%s, sizeof(%s))",
+                                  keyFieldNamePtr, fieldName, fieldName);
+            builder->endOfStatement(true);
+            builder->appendFormat("%s = %s + sizeof(%s)", keyFieldNamePtr,
+                                  keyFieldNamePtr, fieldName);
+            builder->endOfStatement(true);
+        }
+    }
+}
+
+void EBPFTablePSA::emitValueMask(CodeBuilder *builder, const cstring valueMask,
+                                        const cstring nextMask, int tupleId) const {
+    builder->emitIndent();
+    builder->appendFormat("struct %s_mask %s = {0}", valueTypeName, valueMask);
+    builder->endOfStatement(true);
+
+    builder->emitIndent();
+    builder->appendFormat("%s.tuple_id = %s", valueMask, cstring::to_cstring(tupleId));
+    builder->endOfStatement(true);
+    builder->emitIndent();
+    if (nextMask.isNullOrEmpty()) {
+        builder->appendFormat("%s.has_next = 0", valueMask);
+        builder->endOfStatement(true);
+    } else {
+        builder->appendFormat("%s.next_tuple_mask = %s", valueMask, nextMask);
+        builder->endOfStatement(true);
+        builder->emitIndent();
+        builder->appendFormat("%s.has_next = 1", valueMask);
+        builder->endOfStatement(true);
+    }
+}
+
+std::vector<std::vector<const IR::Entry*>> EBPFTablePSA::getConstEntriesGroupedByPrefix() {
+    std::vector<std::vector<const IR::Entry*>> entriesGroupedByPrefix;
+    const IR::EntriesList* entries = table->container->getEntries();
+    if (entries != nullptr) {
```

Comment:
especially for long ifs I find it simpler to `if (!entries) return ...;`
One less level of indentation afterwards.